File: trunk/FarNet/FarNetMan/InputBox.cpp

/*
FarNet plugin for Far Manager
Copyright (c) 2006-2015 Roman Kuzmin
*/

#include "StdAfx.h"
#include "InputBox.h"

namespace FarNet
{;
InputBox::InputBox()
: _maxLength(511)
{}

int InputBox::MaxLength::get()
{
	return _maxLength;
}

void InputBox::MaxLength::set(int value)
{
	if (value < 1)
		throw gcnew ArgumentOutOfRangeException("value");

	_maxLength = value;
}

String^ InputBox::HelpTopic::get()
{
	return _HelpTopic;
}

void InputBox::HelpTopic::set(String^ value)
{
	if (SS(value) && !value->StartsWith("<"))
		throw gcnew ArgumentException("'value' format must be '<FullPath\\>Topic'");

	_HelpTopic = value;
}

bool InputBox::Show()
{
	if (ValueUserScreen::Get()) //_100514_000000
	{
		ValueUserScreen::Set(false);
		Far::Api->UI->SaveUserScreen();
	}

	PIN_ES(pinTitle, Title);
	PIN_ES(pinPrompt, Prompt);
	PIN_ES(pinText, Text);
	PIN_NS(pinHelp, HelpTopic);
	PIN_NS(pinHistory, History);
	CBox box(MaxLength + 1);

	if (!Info.InputBox(&MainGuid, &MainGuid, pinTitle, pinPrompt, pinHistory, pinText, box, box.Size(), pinHelp, Flags()))
		return false;

	Text = gcnew String(box);
	return true;
}

int InputBox::Flags()
{
	int r = 0;
	if (ButtonsAreVisible) r += FIB_BUTTONS;
	if (EmptyEnabled) r += FIB_ENABLEEMPTY;
	if (ExpandEnvironmentVariables) r += FIB_EXPANDENV;
	if (IsPath) r += FIB_EDITPATH;
	if (IsPassword) r += FIB_PASSWORD;
	if (!UseLastHistory) r += FIB_NOUSELASTHISTORY;
	return r;
}

}

File: trunk/FarNet/FarNet/InputBox.cs

/*
FarNet plugin for Far Manager
Copyright (c) 2006-2015 Roman Kuzmin
*/

namespace FarNet
{
	/// <summary>
	/// Input box. It is created by <see cref="IFar.CreateInputBox"/>.
	/// </summary>
	public interface IInputBox
	{
		/// <summary>
		/// Gets or sets the text.
		/// </summary>
		string Text { get; set; }
		/// <summary>
		/// Gets or sets the box title.
		/// </summary>
		string Title { get; set; }
		/// <summary>
		/// Gets or sets the prompt text.
		/// </summary>
		string Prompt { get; set; }
		/// <summary>
		/// Gets or sets the history name.
		/// </summary>
		string History { get; set; }
		/// <summary>
		/// Gets or sets the max text length.
		/// </summary>
		int MaxLength { get; set; }
		/// <summary>
		/// Tells to enable empty input permission.
		/// </summary>
		bool EmptyEnabled { get; set; }
		/// <summary>
		/// Tells that it is used for file system path input.
		/// </summary>
		/// <remarks>
		/// Setting this to true enables some extras, e.g. on typing: a dropdown list of matching available paths.
		/// </remarks>
		bool IsPath { get; set; }
		/// <summary>
		/// Tells that it is used for password input.
		/// </summary>
		/// <remarks>
		/// If it is true then asterisks are displaied instead of input characters.
		/// </remarks>
		bool IsPassword { get; set; }
		/// <include file='doc.xml' path='doc/ExpandEnvironmentVariables/*'/>
		bool ExpandEnvironmentVariables { get; set; }
		/// <include file='doc.xml' path='doc/UseLastHistory/*'/>
		bool UseLastHistory { get; set; }
		/// <summary>
		/// Tells that buttons are visible.
		/// </summary>
		bool ButtonsAreVisible { get; set; }
		/// <summary>
		/// Gets or sets the help topic; the only supported format is "&lt;FullPath\&gt;Topic", see <see cref="IFar.ShowHelp"/>.
		/// </summary>
		string HelpTopic { get; set; }
		/// <summary>
		/// Shows the input box and waits until user press OK or Cancel.
		/// </summary>
		/// <returns>True if OK is pressed.</returns>
		bool Show();
	}
}
